### DISCUSSION TITLE
MuPIF datatypes for musicode schemata

### DISCUSSION THREAD

**Discussion by solidkon** [mupif/mupif#53] [Q&A] [answered]
Hi! I am using the mupif Heavystruct context from the Musicode branch to create a hdf5 from an external file used in simulations.  Since the Heavystruct context uses the schemata, I need to know the datatypes defined there to be able to store my data in the correct types. I am getting an error as I try to pass an array via the services defined in the schema for molecule. 

Specifically, I am trying to pass an array of STR via the following set service:
`atom.getStructure().setBondTypes( np.array(['1', '1', '1']) )`

which raises the error:
`TypeError: No conversion path for dtype: dtype('<U1')`

The corresponding field in schemata is:
```
"structure":{
			"bondNeighbors": { "dtype":"l", "shape":"variable" },
			"bondTypes": { "dtype":"l", "shape":"variable" },
			"name": { "dtype":"a10" },
			"type": { "dtype":"a" },
			"position": { "dtype":"d", "shape":[3],"unit":"AA"},
			"velocity": { "dtype":"d", "shape":[3],"unit":"AA/ps"},
			"residueId": {"dtype":"l" }
		}
```

From a quick search, I found that this is actually an error associated with the hdf5 container and arrays of type str. As you can see from the  passed array I have the values: '1', '1', '1' **_BUT_** regardless of this case, these need to be of type str as they define specific bond types. Is there any workaround for this?

**Comment by eudoxos** [comment]
Hi, for some reason (this is an issue on our side) the documentation about writing the schemata was not showing up in readthedocs, which should be fixed now. Please have a look here: https://mupif.readthedocs.io/en/master/ref.html#mupif.HeavyStruct . Feel free to reopen if you need more clarifications. Good luck!

**Reply by eudoxos** [reply]
This is a MWE iwth your schema which works. No, you cannot save strings where numbers are expected (numpy and hdf5 dtypes pre-define the memory layout).

If you really really need bond types as  strings, I'd suggest using a  single dynamic-size string `"type":"a"` and process the string manually (using `split` and such).

If the number of bonds is significant, define the bonds as a subschema, then for each bond use `choice` field (not yet documented, but looks like this: `"symmetry": { "choice":["none","axial","periodic","translational","rotational"] }`).



```python
schema='''
[{
    "_schema":{
        "name": "test2",
        "version": "1.0"
    },
    "_datasetName":"test2",
    "structure":{
        "bondNeighbors": { "dtype":"l", "shape":"variable" },
        "bondTypes": { "dtype":"l", "shape":"variable" },
        "name": { "dtype":"a10" },
        "type": { "dtype":"a" },
        "position": { "dtype":"d", "shape":[3],"unit":"AA"},
        "velocity": { "dtype":"d", "shape":[3],"unit":"AA/ps"},
        "residueId": {"dtype":"l" }
    }
}]
'''
import numpy as np
import mupif as mp
with mp.HeavyStruct(h5path='/tmp/aaa.h5',mode='overwrite',schemaName='test2',schemasJson=schema) as test:
    test.resize(1)
    test[0].getStructure().setBondTypes(np.array([1,2,3]))
```

**Reply by solidkon** [reply]
Let me be more specific with a more generic example of what these arrays are:

This is a part from a simulation file .mol2, where the bond types are specified:
```
@<TRIPOS>BOND
     1     1     2 1   
     2     1     3 3   
     3     2     4 1   
     4     2     6 2   
     5     4     5 3   
     6     6     7 1   
     7     6     8 1   
     8     7     9 ar  
     9     7    12 ar  
```

The first column is the bond index (just an integer of the bond) the second column is the ith-atom, the third column is the jth-atom (so these two columns describe the connected atoms) and the final column is the bond type. As you can see, these can be numeric value- integers ('1', '2', '3'; _even though it is a type of bond_) or strings ( 'ar' ). As I try to create an hdf5 of a molecule, I have to pass an array describing the bonds. 

So if I consider the first atom, then the array is:
`np.array( ['1', '3' ] )`
**_BUT_** if I consider the seventh atom, then the array is:
`np.array( ['1', 'ar', 'ar'] )`,

where the 'ar' type of bond **can only be described by a string**!

 My problem is what is the type of this array? I want this to be of type string as it covers all possible cases. No operations will be performed that **need integer numbers**. According to the schema, this is: 
`"bondtypes": {"dtype":"l", "shape": "variable"}  `

This is not the same as in bondNeighbors:
`"bondNeighbors": { "dtype":"l", "shape":"variable" }`,

where _integers are expected_. Excuse me for not being entirely clear from the beginning. I hope it makes sense now!

**Reply by eudoxos** [reply]
The first possibility coming to mind is to use the table format of `@<TRIPOS>BOND`, creating a subschema just for bonds. This would be efficient if the number of bonds is in tens or hundereds (there is some overhead for creating a new group). Then there would be just one line per bond, and one value in each column, same as in the `.mol2` file. What are the possible values of bond type? If it is 1, 2, 3 and `ar`, then you could either encode `ar` as `0` (number), or use `"bondType":{ "choice": {'1','2','3','ar'}}` which would encode the values as integers for you (they would be stored as numbers anyway, but you would not see that really).

Note the difference betwen `np.array( ['1','3'])` (array of strings, not supported in HDF5 really) and `np.array([1,3])` (integer array of size 2).

**Reply by solidkon** [reply]
From our discussion I conclude that the best choice is to use your recommendation:
`"bondType":{ "choice": {'1','2','3','ar'}}`, **(1)**

as the possible bond type-values for a generic .mol2 file are:

- 1 = single
- 2 = double
- 3 = triple
- am = amide
- ar = aromatic
- du = dummy
- un = unknown (cannot be determined from the parameter tables)
- nc = not connected

_these are documented in http://chemyang.ccnu.edu.cn/ccb/server/AIMMS/mol2.pdf_
If I understand correctly, the method proposed in **(1)**, makes a pair assignment of an integer to a string described in "choice" key of the schema. If that is the case, then I think it is a good solution. Can you please provide the following:

i) The updated schema field for "bondTypes"?
ii) An example of how to use the service with the new implementation? 

Since the hsf5 does not support arrays of strings I do not know how to pass the information to the hdf5 container.

Thanks a lot again for your prompt responses and support!

**Comment by eudoxos** [comment]
I would store bonds in a separate dataset from atoms, something like this (as I understand the tripos2 mol file format docs, it is only by accident that the benzene molecule has the same number of atoms and bonds, right?):

```json
[
  {
    "_schema": {
      "name": "atoms",
      "ETC just and example for atom": ""
    }
  },
  {
    "_schema": {
      "name": "bonds",
      "version": "1.0"
    },
    "_datasetName": "bonds",
    "ids": { "dtype": "l", "shape": [ 2 ] },
    "type": {
        "choice": [ "none", "single", "double", "triple", "amide", "aromatic", "dummy", "unknown" ] }
  },
  {
    "_schema": {
      "name": "molecule",
      "version": "1.0"
    },
    "_datasetName": "foo",
    "structure": {
      "bonds": {
        "schema": "bonds"
      },
      "atoms": {
        "schema": "atoms"
      }
    }
  }
]
```
and then something like this:
```python
with mp.HeavyStruct(h5path='/tmp/aaa.h5',mode='overwrite',schemaName='molecule',schemasJson=schemata) as mol:
    bonds=mol.getBonds()
    bonds.resize(100)
    bonds[0].setIds([0,1]) # atom ids
    bonds[0].setType("double")
    ...
    atoms=mo.getAtoms()
    atoms.resize(30)
    atoms[0].setBlah(...)
```

**Reply by solidkon** [reply]
Yes in the most general case, the number of atoms is different than the number of unique bonds; unique bonds, in the sense that i-j bond needs to be only described once and .mol2 file only writes the unique bonds with the constraint that the atom ids obey i<j.

In the way you propose to store the bonds we lose the consistency of the array lens since bondedNeighbors have the same len as the bondedTypes (I don't know if this is crucial but it is a nice to have when writing for loops). So for an atom, with an id let's say 0 we have the bondedNeighbors [1, 3, 8] with the associated bondTypes ['1', 'ar', 'am'], describing the bonds 0-1: '1', 0-3: 'ar' and 0-8: 'am'. I also understand that a "lookup" type of implementation also does not work since I will pass the bondTypes as I read them from a .mol2 file (which is ok) BUT when I need the inverse procedure, i.e. to write a .mol2 file from a hdf5 container, I will not have a clue of what the bondType is without an internal db/dict_structure to make the pair key:value association. 

To be more specific, since string arrays are not supported by the hdf5, the lookup would serve to pass a translated array of string types to an integer array and would look like this:

```
"lookup": {
					"1": 1,
					"2": 2,
					"3": 3,
					"ar": 4,
					"am": 5,
					"un": 6,
                                         "nc": 7
				}
```
Which is ambiguous in its semantic representation. Now looking back to your proposal, if I want to create a mol2 and find the bondTypes for each atom, I would need to write a double loop, to generate the pairs  since the bond is described for a specific id pair, right?

```
for iatom in len(molecule.getParts()):
    for jatom in len(molecule.getParts()):
        # some get/set service operations...
        pass
```

**Comment by eudoxos** [comment]
Try this one. No need for double loops. Well, it does not look that pretty but you don't need loops to find the atoms numbers.

```python
schema='''
[
  {
    "_schema": {
      "name": "atoms",
      "version": "1.0"
    },
    "_datasetName": "atoms"
  },
  {
    "_schema": {
      "name": "bonds",
      "version": "1.0"
    },
    "_datasetName": "bonds",
    "ids": { "dtype": "l", "shape": [ 2 ] },
    "type": {
        "choice": [ "none", "single", "double", "triple", "amide", "aromatic", "dummy", "unknown" ] }
  },
  {
    "_schema": {
      "name": "molecule",
      "version": "1.0"
    },
    "_datasetName": "foo",
    "structure": {
      "bonds": {
        "schema": "bonds"
      },
      "atoms": {
        "schema": "atoms"
      }
    },
    "foo":{ "dtype":"l", "must be here":"so that there is some real data for each molecule" }
  }
]
'''
import numpy as np
import mupif as mp
with mp.HeavyStruct(h5path='/tmp/aaa.h5',mode='overwrite',schemaName='molecule',schemasJson=schema) as mols:
    mols.resize(1)
    # atoms for molecule 0
    a0=mols[0].getStructure().getAtoms()
    a0.resize(3)
    a0[0].setElement('H')
    a0[1].setElement('O')
    a0[2].setElement('H')
    # bonds for molecule 0
    b0=mols[0].getStructure().getBonds()
    # create 2 bonds
    b0.resize(2)
    # for each bonds, set ids and type
    b0[0].setIds([0,1])
    b0[0].setType("single")
    b0[1].setIds([1,2])
    b0[1].setType("single")
    # list all bonds where atom #2 is present
    for i in np.where(np.any(b0.getIds()==2,axis=1))[0]:
        print(i,b0[i].getIds(),b0[i].getType())
```
Prints `1 [1 2] single`.

**Reply by solidkon** [reply]
I had another idea that could probably solve this issue related to the hdf5 container and arrays of type string.

I can store the possible bonds in a single string with a delimiter separating them:

e.g.
`bondtypes = '1-2-ar' `
or 
`bondtypes = '1,2,ar'`

In order to have everything well-defined, a corresponding field in  the schema should also define the delimiter. When I want to store the bonds via a set service then I simply write:

`atom.getStructure().setBondTypes( '1-2-3' )`
where I assume that a delimiter = '-' is defined in the schema. In the inverse operation, where I need to get the bondTypes, then I write:

`atom.getStructure().getBondTypes()`
and implicitly a split operation ( bonds.split(delimiter) ) takes care of the returned list/array:
`Out[1]: ['1', '2', 'ar'] `

In this way, the index is retained along the atoms and there is also an index-wise correspondence of the BondedNeighbors array and the BondedTypes list/array. 

Is this a possible solution?

**Comment by eudoxos** [comment, marked as answer]
Hi there, please see 8312af3667bae7ee9f9022e0bbf610cb22dbf9ed for string list in HDF5. An example how to use this is here (dtype `a` or fixed-size string like `a100` and then specify `delim` in addition): https://github.com/mupif/mupif/blob/8312af3667bae7ee9f9022e0bbf610cb22dbf9ed/mupif/tests/test_heavydata.py#L443-L478 . The documentation of the *delim* feature is added to https://mupif.readthedocs.io/en/master/ref.html#mupif.HeavyStruct .

@stanislavsulc please would you merge master to the `Musicode` branch?

**Comment by solidkon** [comment]
I reopen the topic since the returned value of the specified get service is of type str and not of type list/array. Specifically,
I have successfully passed a numpy array of str via the set service but when I need the stored values I get strings for each atom.

I attach the field in the schema where I define this service:

```
"_schema":{
			"name": "musicode.atom",
			"version": "0.5"
		},
		"_datasetName": "atoms",
		
		...

"structure":{
			"bondNeighbors": { "dtype":"l", "shape":"variable" },
			"bondTypes": { "dtype":"a", "delim":"-" },
			"name": { "dtype":"a10" },
			"type": { "dtype":"a" },
			"position": { "dtype":"d", "shape":[3],"unit":"AA"},
			"velocity": { "dtype":"d", "shape":[3],"unit":"AA/ps"},
			"residueId": {"dtype":"l" }
		}
```

I read in the reference manual https://mupif.readthedocs.io/en/master/ref.html#mupif.HeavyStruct that the returned value is of type tuple to prevent in-place item modification. I do not have any problem with this. BUT the returned value of the 

`a.getStructure.getBondTypes()`

should not be of type str since it would imply that further string parsing is required to get the list 

I also attach the part of code I used to get the stored data:

```
with mp.HeavyStruct(h5path=fileh5,mode='readonly',schemaName='musicode.molecule',schemasJson=schemata) as molecules:
        for im, molecule in enumerate(molecules):
            print(molecule.getGeneralInfo().getType())
            atoms = molecule.getParts()
            for iat, atom in enumerate(atoms):
                print(atom.getStructure().getBondTypes())
```

When I execute this I get:

```
['1' '1' '1' '1']
['1' '2' '1']
['2' '1' '1']
['1']
['1']
['1']
['1']
['1']
['1']
```

Even though the data is correct for each atom, the data is of type string (commas are not included as separators of list elements)

**Comment by eudoxos** [comment]
Reading this bit from the e-mail, can you clarify what you need to fixed @solidkon ?

> I have successfully refined the schema according to our needs for the MolecularDynamics API. Additionally, I have defined an internal function get_list() in order to get the needed string array from the .getBondTypes() method. However this needs to be fixed in the future.

**Comment by solidkon** [comment]
With the schema definition defined above, the set service:
 `atom.getStructure().getBondTypes()` 
 
 Returns a string representation of a list, e.g. '['1', '2', '3']'. So I have implemented a temporary function get_list():
 
```
 def get_list(string):
    
    slist = string.strip('[]') 
    new = ''
    for s in slist:
        if s == "'" or s == " ":
            pass
        else:
            new += s
    return new.split(',')
```

To get the BondTypes properly. The get_list receives the string representation and returns its corresponding list object with strings as components. I hope it's clearer now.

**Reply by eudoxos** [reply]
There must be some misunderstanding. `getBondTypes()` will return list of strings, but not as a string but as a tuple of strings; this happens here https://github.com/mupif/mupif/blob/6d4462d5f07a045568e5b030073fb0bd71aa9409/mupif/heavystruct.py#L489

Are you not converting the tuple to string first by yourself, accidentally? I have the intent to understand what you do and make your experience less painful. Can you perhaps paste some MWE showing the issue you are trying to solve?

Just a note, if you *really* need to parse string representation, use [ast.literal_eval](https://docs.python.org/3.8/library/ast.html#ast.literal_eval) (for safe evaluation) or [eval](https://docs.python.org/3.8/library/functions.html#eval) (includes execution of code in the string, if any).

**Reply by eudoxos** [reply]
PS just as an exercise, this might be a shorter version of your `get_list`:
```python
def get_list(string): return string.translate(None,"[]' ").split(',')
```

**Reply by solidkon** [reply]
I attach a .zip folder with a MWE of what I get. 
[MWE.zip](https://github.com/mupif/mupif/files/8419810/MWE.zip)

**Reply by eudoxos** [reply]
Thanks, this is helpful. The data stored in the HDF5 file are strange already; according to the schema, they should be `"1-1-1-1"` (string delimited by `-`, you must have stored a string instead of a link when calling `setBondTypes` already. The call should be `setBondTypes([1,1,1,1])`, for instance. Can you confirm that the problem disappears when you fix the setter?

![Screenshot_20220405_201040](https://user-images.githubusercontent.com/1029876/161821948-3b098246-d1dc-430b-9997-78fb74675d25.png)

**Reply by solidkon** [reply]
Well, actually an example of a set service used for this one is:
`atom.getStructure().setBondTypes( ['1', '1', '1', '1'])`

It should generate the same issue with the container

**Reply by eudoxos** [reply]
You have to send me the generating code as well. I tried this and don't see your issue:

```python
# -*- coding: utf-8 -*-
"""
Created on Tue Apr  5 18:49:30 2022

@author: Kostis
"""
import mupif as mp

gstarhdf5 = gstarhdf5 = 'Gstar.h5'

with mp.HeavyStruct(h5path=gstarhdf5, mode='overwrite',schemaName='musicode.grain',schemasJson=open('schemata_4.json').read()) as grainStar:
    grainStar.resize(1)
    for g in grainStar:
        molecules=g.getParts()
        molecules.resize(2)
        for im,mol in enumerate(molecules):
            atoms=mol.getParts()
            atoms.resize(10)
            for ia,atom in enumerate(atoms):
                atom.getStructure().setBondTypes(['1','2','3','4','foo','bar'])

with mp.HeavyStruct(h5path=gstarhdf5, mode='readwrite', schemaName='musicode.grain') as grainStar:
    for g in grainStar:
       molecules = g.getParts()
       for im,mol in enumerate(molecules):
           atoms = mol.getParts()
           for ia,atom in enumerate(atoms):
               print(type(atom.getStructure().getBondTypes()))
               print(atom.getStructure().getBondTypes())

```

**Reply by solidkon** [reply]
I now attach the generated code for the Gstar.h5. I have an external module (grainstar_functions.py) with the initialization and the update functions. The information comes from a .mol2 file, also attached in the zipped directory.
[Not_so_MWE.zip](https://github.com/mupif/mupif/files/8424740/Not_so_MWE.zip)
.

**Reply by eudoxos** [reply]
I run you Not_so_MWE.zip and the return value of `getBondTypes()` is correct (e.g. the tuple `('1', '1', '1', '1')`, not its string representation) and the inside of the HDF5 file looks also good:
![Screenshot_20220406_095717](https://user-images.githubusercontent.com/1029876/161924987-5c562537-436a-4dd0-aa7a-ec9467f6943b.png)
Is the issue solved then?

**Reply by solidkon** [reply]
Dear eudoxos... Is there a possibility that I do not have the latest implementation properly installed on my pc (?!) 

I probably need to do a full uninstall of the mupif repository and redownload it again, just to be sure. After the latest implementation, I performed a 

`git pull `

command inside the repository. Is this not correct?

**Reply by eudoxos** [reply]
It is true that master branch is a few commits ahead, but the relevant bits should have been merged to Musicode branch already. Perhaps your local repo is not updated? What is the output of
```
python3 -c 'import mupif; print(mupif.util.getVersion())'
```
(you might need to install `gitpython` first). I see now that a previous version (which would ignore `delim`) would convert the object to string and store it (something to fix perhaps so that such auto-conversions don't happen). The installation is here https://mupif.readthedocs.io/en/latest/userman.html#full-source .

You can double-check by running MWE on your machine, inspecting the H5 file with `silx view Gstart.h5` (`pip3 install silx` if needed) — when you open the hierarchy all the way to the atom, you will see what it looks like. Here it is like this:

![Screenshot_20220406_100314](https://user-images.githubusercontent.com/1029876/161926108-6cff463c-abb0-4968-a08c-6e41340962ff.png)

**Reply by eudoxos** [reply]
> a
> 
> `git pull `
> 
> command inside the repository. Is this not correct?

It is probably not sufficient - the repo is not the installation itself. You might try running `pip install -e .` which will just symlink the repo to the python module path, and then just pulling the repo should be enough.

**Reply by solidkon** [reply]
No the reinstallation procedure does not work. Still the import brings an older version of mupif which handles the hdf5 incorrectly. Can you please provide instructions of a full mupif uninstall via anaconda? (both for administration privileges and for user?)

**Reply by solidkon** [reply]
After a few attempts I've been able to successfully install it. Now it works as expected, returning a tuple instead of a string. Thank you for your support! I mark this topic as answered
